### DISCUSSION TITLE
Can't interact with DataTable after loading state in macOS Terminal.app

### DISCUSSION THREAD

**Discussion by frankie567** [Textualize/textual#3911] [Q&A] [answered]
When mounting a `DataTable`, it can directly be interacted with using arrow keys on the keyboard (I suppose it gets automatic focus).

However, this behavior breaks under macOS Terminal.app, after switching the [`loading` state](https://textual.textualize.io/guide/widgets/#loading-indicator). When `loading` is set to `False`, the table appears but the arrow keys have no effect.

With WezTerm, keyboard input is also inactive after loading, but since it detects the mouse cursor (contrary to Terminal.app), the "focus" can be revived.

## Reproducible example

```py
from asyncio import sleep
from random import randint

from textual import work
from textual.app import App, ComposeResult
from textual.widgets import DataTable

ROWS = [
    ("lane", "swimmer", "country", "time"),
    (4, "Joseph Schooling", "Singapore", 50.39),
    (2, "Michael Phelps", "United States", 51.14),
    (5, "Chad le Clos", "South Africa", 51.14),
    (6, "László Cseh", "Hungary", 51.14),
    (3, "Li Zhuhao", "China", 51.26),
    (8, "Mehdy Metella", "France", 51.58),
    (7, "Tom Shields", "United States", 51.73),
    (1, "Aleksandr Sadovnikov", "Russia", 51.84),
    (10, "Darren Burns", "Scotland", 51.84),
]


class DataApp(App):
    CSS = """
    DataTable {
        height: 1fr;
    }
    """

    def compose(self) -> ComposeResult:
        yield DataTable()

    def on_mount(self) -> None:
        data_table = self.query_one(DataTable)
        data_table.loading = True
        self.load_data()

    @work
    async def load_data(self) -> None:
        await sleep(randint(2, 10))
        data_table = self.query_one(DataTable)
        data_table.add_columns(*ROWS[0])
        data_table.add_rows(ROWS[1:])
        data_table.loading = False


if __name__ == "__main__":
    app = DataApp()
    app.run()
```

# Textual Diagnostics

## Versions

| Name    | Value  |
|---------|--------|
| Textual | 0.46.0 |
| Rich    | 13.7.0 |

## Python

| Name           | Value                                                   |
|----------------|---------------------------------------------------------|
| Version        | 3.12.0                                                  |
| Implementation | CPython                                                 |
| Compiler       | Clang 15.0.0 (clang-1500.0.40.1)                        |
| Executable     | /Users/fvoron/Development/polar/server/.venv/bin/python |

## Operating System

| Name    | Value                                                                                                 |
|---------|-------------------------------------------------------------------------------------------------------|
| System  | Darwin                                                                                                |
| Release | 23.1.0                                                                                                |
| Version | Darwin Kernel Version 23.1.0: Mon Oct  9 21:27:24 PDT 2023; root:xnu-10002.41.9~6/RELEASE_ARM64_T6000 |

## Terminal

| Name                 | Value                |
|----------------------|----------------------|
| Terminal Application | Apple_Terminal (452) |
| TERM                 | xterm-256color       |
| COLORTERM            | *Not set*            |
| FORCE_COLOR          | *Not set*            |
| NO_COLOR             | *Not set*            |

## Rich Console options

| Name           | Value                |
|----------------|----------------------|
| size           | width=245, height=62 |
| legacy_windows | False                |
| min_width      | 1                    |
| max_width      | 245                  |
| is_terminal    | True                 |
| encoding       | utf-8                |
| max_height     | 62                   |
| justify        | None                 |
| overflow       | None                 |
| no_wrap        | False                |
| highlight      | None                 |
| markup         | None                 |
| height         | None                 |

**Comment by davep** [comment, marked as answer]
You should find that the behaviour you're seeing is the same in any environment; not in some particular terminal or other. The issue here is that you're disabling the only focusable widget in your screen, so once doing that nothing has focus. When you set `loading` to `False` again focus won't be "stolen" (if there was more than one focusable widget on the screen you don't want focus pulled away from it just because the loading status of another widget has changed).

In your code here, you want to be adding `data_table.focus()` after you've set `data_table.loading = False`.

**Reply by frankie567** [reply]
Ah, right! Had the feeling it had to do with focus. Thanks for the detailed explanation and fix!

What was a bit surprising is, without using the loading state, it *"just works"*: there is a focus on the widget. With loading, and without manually calling `.focus()`, we end up in a state where we're kinda stuck (at least in Terminal.app). Might be worth to mention in the loading example: https://textual.textualize.io/guide/widgets/#loading-indicator

**Reply by davep** [reply]
Do note that this is normal focus behaviour for Textual applications; you can focus nothing. This is also what would happen if you had a screen filled with widgets that can't receive focus. Alternatively, if you were to add some widget that can receive focus, such that it comes after your `DataTable`, focus would have moved to it when you set `loading` to `True`.

I guess what I'm saying there is that while it could perhaps merit a mention in the docs there, there's a handful of different cases that would need to be documented.
